From b1d2ad484993c2f170e53228c639502f92bf93eb Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Tue, 19 Oct 2021 11:16:33 +0300
Subject: [PATCH] [SW-56231] boot_desc: update checksum logic

Based on the firmware descriptor version we choose the checksum
field. For version <3 sha256 field is considered as valid checksum
algorithm, and for version >=3 sha384 field is considered as valid
checksum algorithm.

Please check https://gerrit.habana-labs.com/#/c/138267/ for detailed
background.

Change-Id: Ic3946a65f3362879f8db670d690190880314d21e
---

diff --git a/lib/common/boot_desc.c b/lib/common/boot_desc.c
index 8cbc974..1ce014f 100644
--- a/lib/common/boot_desc.c
+++ b/lib/common/boot_desc.c
@@ -87,6 +87,7 @@
 	return count;
 }
 
+#define MULTI_CSUM_BOOT_VER	3
 u32 boot_desc_get_img_desc(enum fw_post_reset_cmd upd_status,
 			   enum hl_fw_img_type type, struct boot_desc *desc)
 {
@@ -107,8 +108,12 @@
 		desc->loadaddr = ((u64)img_desc[i].load_high[idx] << 32) |
 					img_desc[i].load_low[idx];
 		desc->desc_ver = img_desc[i].desc_ver;
-		memcpy(desc->chksum, img_desc[i].chksum[idx],
-		       ARRAY_SIZE(desc->chksum));
+		if (desc->desc_ver >= MULTI_CSUM_BOOT_VER)
+			memcpy(desc->chksum, img_desc[i].chksum2[idx],
+			       ARRAY_SIZE(desc->chksum));
+		else
+			memcpy(desc->chksum, img_desc[i].chksum[idx],
+			       ARRAY_SIZE(desc->chksum));
 
 		return idx;
 	}
